<div class="dialogcontainer">
	<h2 class="bottommargin1">CpGAT Region Tool: Parameters</h2>
	<p>The default parameters and options are optimized for annotating the current genome region using previously-computed spliced alignments. Only the <b>species model</b> may need to be adjusted for your taxonomic group. </p>
	<p>Additional <b> Parameters and options</b> can be selected according to your needs:</p> 
	<ul class="bullet1 indent2">
		<li>Custom genome sequence: </li>
		<li>Custom protein/transcript files:</li>
		<li>Repeat Mask File:</li>
		<li>Species Models:</li>
		<li>Repeat Mask File:Other Options
			<ul class="bullet1">
				<li>Skip Mask</li>
				<li>Relax UniRef</li>
				<li>Skip PASA</li>
				<li>Skip BLAST</li>
			</ul>
		<li>Select Reference Protein Dataset:</li>
		<li>Repeat Mask File:</li>
	</ul>
<p>Additional parameters can be modified from the configuration file found at <span class="plaintext">/xGDBvm/src/CpGAT/CpGAT.conf</span>:</p>
<ul class="bullet1 indent2">
	<li></li>
	<li></li>
	<li></li> 
</ul>

<p>See also <a href="http://goblinx.soic.indiana.edu/wiki/doku.php?id=cpgat">xGDBvm wiki</a> for more information about CpGAT.</p>

			<span class="heading normalfont linkback"> <a class="smallerfont" href="/XGDB/help/cpgat.php#cpgat_tool_parameters">View this in Help Context</a> (cpgat.php/cpgat_tool_parameters)</span>
</div>
