<div class="dialogcontainer">

	<h2 class="bottommmargin1">What is GAEVAL?</h2>
<hr class="featuredivider" />
		<p>GAEVAL (Gene Annotation EVALuation) is system for comparing gene models with trancript data,  based on splice-aligned transcript evidence overlapping with a locus. It flags and categorizes incongruencies, and also identifies potential alternative splicing. A key feature is the GAEVAL <b>Quality Score</b>, a measure of the degree to which the available evidence supports a given model.</p>

<p>GAEVAL scripts are run automatically each time gene model or transcript information are created or updated. Both gff-derived and CpGAT-derived gene models are evaluated, and the results can be viewed in several ways:</p>

				<ul class="bullet1">
				<li>In the Genome Context display as a glyph above each model: <b>right click</b> on trackname to turn on GAEVAL flags)</li>
				<li>In a GAEVAL Table (Analysis of Transcripts): Choose <b>'Gene Quality - GAEVAL'</b> from left GDB menu</li>
				<li>In a Gene/Locus table: Choose <b>'Gene Loci - from gff'</b> or <b>'Gene Loci - from CpGAT'</b> from the GDB left menu.</li>
				<li>A comprehensive <b>GAEVAL record</b> is available for each gene model by clicking the GAEVAL glyph mentioned above, or from the GAEVAL table.</li>
				</ul>

<p>See also <b><a href="/src/GAEVAL/docs/index.html">detailed GAEVAL documentation</a></b>, which also includes a guide for those wanting to implement it locally.
</p>

			<span class="heading normalfont linkback"> <a class="smallerfont" href="/XGDB/help/gaeval.php#gaeval_overview">View this in Help Context</a> (gaeval.php/gaeval_overview)</span>

</div>
