
<div class="dialogcontainer">
	<h2>Gene Model Track (Genome Context View)</h2>

<h3 class="topmargin1 bottommargin1">Gene Models (from .gff3)</h3>

<img alt="?" src="/XGDB/help/images/genemodel_glyph_example.png" />

<p>Annotated Gene Models represent the predicted structure of transcript and
coding regions, and may include alternative splicing isoforms.</p>
<ul class="bullet1">
    <li>Exons: solid blocks</li>
    <li>Introns: narrow lines</li>
    <li>Coding region Start/Stop:  Green/Red triangles</li>
</ul>
     <p>Click on any image to view its Annotation Record</p>

     <p>Use the track menu -> "Track Options" to show GAEVAL gene quality flags.</p>

<p>This track displays gene models computed elsewhere and uploaded to xGDBvm 
via  a .gff3- formatted data table.


<h3 class="topmargin1 bottommargin1">Gene Models (from CpGAT)</h3>

<img alt="?" src="/XGDB/help/images/genemodel_cpgat_glyph_example.png" />

<p>This track represents a gene model created using the CpGAT tool in xGDBvm.
(Alternatively, it may be computed elsewhere and uploaded to xGDBvm.)

<p>CpGAT uses a combination of spliced alignment evidence and ab initio gene
finder data to create the most likely model(s) in a genomic region. 
See Help -> CpGAT help for more information.</p>

<p>CpGAT gene models can be generated on a region basis using the tool accessed
 from the  Genome Context View submenu (Analysis Tools -> CpGAT).</p>
 
<p>The record was generated via the yrGATE tool and was published to the genome browser. It may confirm, improve, or reject a published model on
which it's based, depending on Annotation Class (color coded).</p>

<span class="heading normalfont linkback"> <a class="smallerfont" href="/XGDB/help/yrgate.php#yrgate_track_info">View this in Help Context (yrgate/yrgate_track_info)</a></span>

</div>
