<div class="dialogcontainer">
	<h2 class="bottmmargin1">Genome Browser </h2>

<p>The xGDBvm Genome Browser uses data stored in MySQL tables to generate a 2_D view of each genome segment and its aligned features.</p>

<p>Use the Genome Browser to browse, search, analyze, or annotate a genome region or its features.</p>

<hr class="featuredivider" />
	<h3 class="topmargin1">Genome Context View</h3>
	<ul class="bullet1 ">
		<li>The Genome Context View displays all features for a genome region as glyphs that are color-coded as to data type.</li>
		<li>To view a region in Genome Context, you simply specify a genome segment and a starting and ending coordinate.</li>
		<li>You can zoom in / zoom out / scroll along the genome using region controls, or view region at the DNA level.</li>
		<li>Clicking on a track feature brings up more information about that sequence or feature.</li>
		<li>You can add a custom track based on genome data formatted in the GFF3 format.</li>
		<li>You can re-annotate a region using the CpGAT tool</li>
		<li>You can manually annotate a region using the yrGATE tool</li>
	</ul>


<hr class="featuredivider" />
	<h3>Search Tools</h3>
	<ul class="bullet1">
		<li>ID/Keyword (Advanced) Search), accessed from the left menu, allows retrieval of sequence data or upstream/downstream/intron/exon sequence.</li>
	</ul>

<hr class="featuredivider" />
	<h3>Analysis Tools</h3>
	<ul class="bullet1">
		<li>
			When viewing a region in Genome Context View, accessing a tool automatically loads the genome region being viewed as a query sequence.
		</li>
		<li>Analysis tools include Blast, GenomeThreader, Download</li>
		<li>
			Tools can also be accessed from the left menu.
		</li>
	</ul>

<hr class="featuredivider" />
	<h3>Annotation Tools</h3>
	<ul class="bullet1">
		<li>the yrGATE tool allows users to manually annotated a genic region to amend or correct existing annotations.</li>
		<li>
			CpGAT provides the user with a customizable, region-specific annotation pipeline, similar to the whole genome version used in GDB creation. Output can be loaded as a separate track.
		</li>
	</ul>

<hr class="featuredivider" />
	<h3>Annotation Tables</h3>
	<ul class="bullet1">
		<li>
			Annotation tables display each type of annotation or aligned sequence in linear order on each genome segment
		</li>
		<li>Tables are a convenient way to search and download data of interest</li>
		<li>Tables also integrate GAEVAL gene quality data which provide a measure of annotation congruency with alignment evidence.</li>
	</ul>


			<span class="heading normalfont linkback"> <a class="smallerfont" href="/XGDB/help/using_gdb.php#genome_overview">View this in Help Context</a> (using_gdb.php/genome_overview)</span>

</div>
