<div class="dialogcontainer">

	<h1 class="bottommmargin1">Loci for this sequence<span class="heading"> &nbsp;&nbsp;&nbsp; See also <a title="Genome Browser Help" href="/help/xgdb.php">Help Pages</a></span></h1>
		<hr class="featuredivider" />

	<p class="topmargin1">This table lists all locations in the genome where this sequence was splice-aligned with above-threshold quality scores. The <b>currently viewed location</b> is show with a dark cell border.</p>
	
	<p>To view a <b>different entry's Record Page</b>, click on its Entry number in column 1</p>
	
	<p>To view an <b>entry in Genome Context</b>, click on the desired "chr" or "gi" in column 2. </p>
	
	<p>Table Columns:</p>
	<ul class="bullet1">
		<li><b>Entry</b>: For EST, cDNA, protein, more than one alignment to the genome may pass quality threshold. IF so, each alignment is assigned an entry and is listed in this table.</li>
		<li><b>chr / gi</b>: the ID of the genomic segment</li>
		<li><b>Strand</b>: + or - strand </li>
		<li><b>Left/Right</b>: left - right coordinates that align to the genome (may or may not correspond to sequence ends)</li>
		<li><b>Sim</b>: fraction of exons with similarity to genome sequence, as calculated by the spliced alignment program.</li>
		<li><b>Cov</b>: fractional amount of the exons covered by alignment to genome</li>
		<li><b>Structure</b>: gene structure diagram. NOTE: for EST (red) and cDNA (blue) alignments, non-cognate alignments (having less than maximal simiarity/coverage scores) will be shown in pink and gray, respectively.</li>
		<li><b>Splice Site Distribution</b>: the pattern of splice site distribution for each spliced alignment</li>
	</ul>

</div>
