
<div class="dialogcontainer">
	<h2>Genome Sequence Edits</h2>

	<p>If you have detected one or more errors in the genome sequence, you may edit the genome sequence using this tool. The edits will affect this gene model only and have no effect on other sequence alignments or annotations.</p>
	<p>Click to open the edit tool and follow the instructions below:</p>
	<p><b>How to enter a genome edit:</b></p>
	<ol class="bullet1">
		<li>Open the Genome Edit tool</li>
		<li>Scroll to the region to be edited</li>
		<li>Select Type of Edit (deletion; change; insertion)</li>
		<li> Enter bp Position</li>
		<li>If insertion or change, enter New Base(s)</li>
		<li>Click "Add New Edit"</li>
		<li>To add more edits, repeat steps 2-5 in a new entry row.</li>
		<li>Click "Add Edits to Annotation Record". The edit data now appears in the Genome Sequence Edits window</li>
		<li>Click to close the Genome Sequence Editor. </li>
		<li>Now, re-run the ORF finder to view the ORF prediction based on editing sequence.</li>
	</ol>

	<p><b>ANNOTATION TIP:</b> If you suspect genome sequence errors may be preventing the identification of a complete ORF, use the "Sequence Level View" from Genome Context View (accessed under the submenu under "View"...), and compare genome sequence to aligned sequence in the exonic regions of this locus. Possible genome errors will appear as indels between genomic sequence and aligned transcript sequence.<img class="helpfigure" src="/help/images/anno/sequence_error.png" style="width:40%" alt=" sequence error screenshot" /></p>

	<p>See <a title="Genome Annotation Help - Annotation" href="/help/anno.php" target="_blank">Annotation Help</a> for more details, including a video tutorial.</p>

</div>

